File: lib/TemplateStore.swift
import Foundation

enum TemplateStoreError: Error {
    case fileNotFound(String)
    case decodeFailed(Error)
}

final class TemplateStore {
    // Maps intentKey -> [template strings]
    private(set) var templates: [String: [String]]

    init(jsonFileName: String = "templates", bundle: Bundle = .main) throws {
        guard let url = bundle.url(forResource: jsonFileName, withExtension: "json") else {
            throw TemplateStoreError.fileNotFound("\(jsonFileName).json")
        }

        do {
            let data = try Data(contentsOf: url)
            self.templates = try JSONDecoder().decode([String: [String]].self, from: data)
        } catch {
            throw TemplateStoreError.decodeFailed(error)
        }
    }

    /// Returns a random template for an intentKey, optionally preferring ones whose required slots exist.
    func randomTemplate(
        for intentKey: String,
        slots: [String: String],
        fallbackKey: String = "unknown"
    ) -> String {
        let list = templates[intentKey] ?? templates[fallbackKey] ?? ["Sorry—can you rephrase that?"]

        // Prefer templates that can be fully filled (all placeholders present in slots)
        let viable = list.filter { template in
            missingPlaceholders(in: template, slots: slots).isEmpty
        }

        return (viable.randomElement() ?? list.randomElement())!
    }

    private func missingPlaceholders(in template: String, slots: [String: String]) -> [String] {
        // finds {PLACEHOLDER}
        let pattern = #"\{([A-Z_]+)\}"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }

        let ns = template as NSString
        let matches = regex.matches(in: template, range: NSRange(location: 0, length: ns.length))

        var missing: [String] = []
        for m in matches {
            let key = ns.substring(with: m.range(at: 1))
            if slots[key] == nil { missing.append(key) }
        }
        return missing
    }
}

// Simple placeholder renderer: replaces {KEY} with slots["KEY"] if present.
enum TemplateRenderer {
    static func render(_ template: String, slots: [String: String]) -> String {
        let pattern = #"\{([A-Z_]+)\}"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return template }

        let ns = template as NSString
        let matches = regex.matches(in: template, range: NSRange(location: 0, length: ns.length))

        var result = template
        for m in matches.reversed() {
            let key = ns.substring(with: m.range(at: 1))
            guard let value = slots[key],
                  let r = Range(m.range(at: 0), in: result) else { continue }
            result.replaceSubrange(r, with: value)
        }
        return result
    }
}

File: lib/template_store.dart
import 'dart:convert';
import 'package:flutter/services.dart' show rootBundle;

class TemplateStore {
  final Map<String, List<String>> templates;

  TemplateStore(this.templates);

  static Future<TemplateStore> loadFromAsset(String path) async {
    final jsonStr = await rootBundle.loadString(path);
    final decoded = json.decode(jsonStr);
    if (decoded is! Map) {
      throw Exception('templates.json must be a JSON object of intentKey -> [templates]');
    }

    final Map<String, List<String>> casted = {};
    decoded.forEach((key, value) {
      if (value is List) {
        casted[key.toString()] = value.map((e) => e.toString()).toList();
      }
    });

    return TemplateStore(casted);
  }

  /// Pick a template with priority:
  /// - if NOUN exists, prefer templates containing {NOUN}
  /// - if PLACE exists, prefer templates containing {PLACE}
  /// - always prefer templates whose placeholders are all fillable
  /// - deterministic: pick the first match (no randomness)
  String pickTemplate(String intentKey, Map<String, String> slots,
      {String fallbackKey = 'unknown'}) {
    final list = templates[intentKey] ?? templates[fallbackKey] ?? const ["Sorry—can you rephrase that?"];

    bool fillable(String t) => _missingPlaceholders(t, slots).isEmpty;

    final viable = list.where(fillable).toList();
    if (viable.isEmpty) return list.first;

    // Priority: NOUN templates first when noun exists
    if (slots.containsKey('NOUN')) {
      final nounFirst = viable.where((t) => t.contains('{NOUN}')).toList();
      if (nounFirst.isNotEmpty) return nounFirst.first;
    }

    // Priority: PLACE templates first when place exists
    if (slots.containsKey('PLACE')) {
      final placeFirst = viable.where((t) => t.contains('{PLACE}')).toList();
      if (placeFirst.isNotEmpty) return placeFirst.first;
    }

    return viable.first;
  }

  static List<String> _missingPlaceholders(String template, Map<String, String> slots) {
    final re = RegExp(r'\{([A-Z_]+)\}');
    final missing = <String>[];
    for (final m in re.allMatches(template)) {
      final key = m.group(1);
      if (key != null && !slots.containsKey(key)) missing.add(key);
    }
    return missing;
  }
}

String renderTemplate(String template, Map<String, String> slots) {
  return template.replaceAllMapped(RegExp(r'\{([A-Z_]+)\}'), (m) {
    final key = m.group(1);
    if (key == null) return m.group(0) ?? '';
    return slots[key] ?? m.group(0) ?? '';
  });
}